### PR TITLE
Multiple samples - Add done loading listeners to avoid silent failure

### PR DESCRIPTION
**Pull request by JonLavi**
We've recently had a lot of customers pointing out troubles with loadables, which is why we need to highlight this pattern a bit more. I'm also going to piggyback the issue regarding garbage collection and make some of the changes here where I've already had the sample open.

* Made some of the loadables class members, to avoid garbage collection

* Added checks for LoadStatus.LOADED, either showing a message if failed to load, or enabling the rest of the sample's logic.
	Note that we're triggering the layer's loading process by adding it to the map _outside_ of the doneLoadingListener, and just using the latter for error reporting or enabling functionality. The alternative would have been to add the layers to the map inside the listener, and start the loading with `loadAsync()`. However that can be a bit confusing and shouldn't be the pattern we promote. 

* in a couple of samples I've removed service URL class members, and just put the URLs in-line when the appropriate table/layer is created

cc @mbcoder

### CODE REVIEW

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java [hunk 1]
```diff
@@ -18,19 +18,23 @@
...
 import com.esri.arcgisruntime.layers.FeatureLayer;
+import com.esri.arcgisruntime.loadable.LoadStatus;
 import com.esri.arcgisruntime.mapping.ArcGISMap;
 import com.esri.arcgisruntime.mapping.Basemap;
 import com.esri.arcgisruntime.mapping.view.MapView;
 
 public class DisplayAnnotationSample extends Application {
 
     private MapView mapView;
+    private FeatureLayer riverFeatureLayer;
```
Comment (Rachael-E) [inline display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java]: just a heads up from https://github.com/Esri/arcgis-runtime-samples-java/pull/511/files which is still to be approved: when putting loadables as member variables to protect against garbage collection, I was adding the following comment to make it clear why we did this. This comment may or may not be agreed on but wanted to doc it here if it's approved so we can add the same comment here for the loadables.
```suggestion
// keeps loadables in scope to avoid garbage collection
    private FeatureLayer riverFeatureLayer;
```

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java [hunk 2]
```diff
@@ -54,19 +58,35 @@ public void start(Stage stage) {
...
 
             // create an annotation layer from a feature service
-            AnnotationLayer annotationLayer = new AnnotationLayer("https://sampleserver6.arcgisonline.com/arcgis/rest/services/RiversAnnotation/FeatureServer/0");
+            annotationLayer = new AnnotationLayer("https://sampleserver6.arcgisonline.com/arcgis/rest/services/RiversAnnotation/FeatureServer/0");
 
             // add the annotation layer to the map
             map.getOperationalLayers().add(annotationLayer);
 
+            // check that the feature layer has loaded
+            // this event gets triggered asynchronously when the feature layer that has been added to the map is loaded
```
Comment (Rachael-E) [inline display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java]: I'm thinking a slight rewording of the comments here and throughout the PR, what do you think?
```suggestion
            // check that the feature layer has loaded by adding a done loading listener (triggered asynchronously)
            // check for the load status of the layer and if it hasn't loaded, report an error
```

File: image_layers/map-image-layer-sublayer-visibility/src/main/java/com/esri/samples/map_image_layer_sublayer_visibility/MapImageLayerSublayerVisibilitySample.java [hunk 3]
```diff
@@ -75,27 +74,38 @@ public void start(Stage stage) {
       controlsVBox.getChildren().addAll(citiesBox, continentsBox, worldBox);
       controlsVBox.getChildren().forEach(c -> ((CheckBox) c).setSelected(true));
 
-      // create a ArcGISMap with the a BasemapType Topographic
+      // create a map view
+      mapView = new MapView();
+
+      // create a ArcGISMap with the a BasemapType Topographic and set it to the map view
```
Comment (Rachael-E) [inline image_layers/map-image-layer-sublayer-visibility/src/main/java/com/esri/samples/map_image_layer_sublayer_visibility/MapImageLayerSublayerVisibilitySample.java]: we can "plain english" this comment a little more
```suggestion
      // create an ArcGISMap with the topographic basemap and set it to the map view
```

**Review by Rachael-E — APPROVED** [review]
thanks @JonLavi looks good. I've put in some comment suggestions, take or leave them as you like!

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java (hunk 1 above)
Comment (JonLavi) [reply display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java]: Definitely need to add the same comment in these samples! 👍

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java (hunk 2 above)
Comment (JonLavi) [reply display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java]: I'm having some problems with this wording, because to me it reads as if the action of adding the listener is what's triggered asynchronously. Also the repetition of 'check' in the beginning bugs me a bit. How about:
```suggestion
            // add a done loading listener, with a runnable that is triggered asynchronously when the feature layer has loaded
            // check for the load status of the layer and if it hasn't loaded, report an error
```
A bit wordy, but hopefully clearer?

**Comment by JonLavi** [comment]
Thanks for the comments @Rachael-E ! I've made the Layers that are added to OperationalLayers local variables again, following Colins advice in your PR, that these won't get garbage collected. Let me know what you think about the reworded comment for the done loading listener?

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java (hunk 2 above)
Comment (Rachael-E) [reply display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java]: @JonLavi yeh I like this wording! Minor change I'd make would be to replace "gets" with "is", just reads a little better I find.

File: editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java [hunk 4]
```diff
@@ -53,11 +53,10 @@
 
 public class UpdateAttributesSample extends Application {
 
+  private ArcGISFeature selected;
```
Comment (Rachael-E) [inline editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java]: with your comment from my garbage collection PR in mind, should we change this to `selectedFeature`?

**Review by Rachael-E — APPROVED** [review]
thanks @JonLavi, looks good! I left a comment in response to re-wording and a note here about changing variable name from `selected` to `selectedFeature`.

File: editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java (hunk 4 above)
Comment (JonLavi) [reply editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java]: Definitely! Thanks for pointing that out!

**Comment by JonLavi** [comment]
Thanks for the review @Rachael-E !
@tschie could you review when you have some time? Thanks!

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java [hunk 5]
```diff
@@ -65,8 +67,25 @@ public void start(Stage stage) {
             // add the annotation layer to the map
             map.getOperationalLayers().add(annotationLayer);
 
+            // add a done loading listener, with a runnable that gets triggered asynchronously when the feature layer has loaded
+            // check for the load status of the layer and if it hasn't loaded, report an error
```
Comment (tschie) [inline display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java]: too verbose

```suggestion
            // show alert if layer fails to load
```

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java [hunk 6]
```diff
@@ -65,8 +67,25 @@ public void start(Stage stage) {
...
+            riverFeatureLayer.addDoneLoadingListener(() -> {
+                if (riverFeatureLayer.getLoadStatus() != LoadStatus.LOADED) {
```
Comment (tschie) [inline display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java:72]: This is okay, but seems a little overkill for a simple sample. We've had discussions before about omitting some best practices for the sake of brevity/focus. Should get approval from Lucas/Mike before making more of these changes.

File: display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java [hunk 7]
```diff
@@ -65,8 +67,25 @@ public void start(Stage stage) {
...
+                    new Alert(Alert.AlertType.ERROR, "Error loading Feature Layer.").show();
+                }
+            });
+
+            // add a done loading listener, with a runnable that gets triggered asynchronously when the feature layer has loaded
+            // check for the load status of the layer and if it hasn't loaded, report an error
+            annotationLayer.addDoneLoadingListener(()->{
```
Comment (tschie) [inline display_information/display-annotation/src/main/java/com/esri/samples/display_annotation/DisplayAnnotationSample.java]:
```suggestion
            annotationLayer.addDoneLoadingListener(() -> {
```

File: editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java [hunk 8]
```diff
@@ -118,21 +114,30 @@ public void start(Stage stage) {
...
+      ArcGISMap map = new ArcGISMap(Basemap.Type.STREETS, 40, -95, 4);
+      mapView.setMap(map);
+
       // create service feature table from URL
-      featureTable = new ServiceFeatureTable(FEATURE_LAYER_URL);
+      featureTable = new ServiceFeatureTable("https://sampleserver6.arcgisonline.com/arcgis/rest/services/DamageAssessment/FeatureServer/0");
 
-      // create a feature layer from table
-      featureLayer = new FeatureLayer(featureTable);
+      // create a feature layer from service feature table
```
Comment (tschie) [inline editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java]:
```suggestion
      // create a feature layer with the feature table
```

File: editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java [hunk 9]
```diff
@@ -118,21 +114,30 @@ public void start(Stage stage) {
...
+      FeatureLayer featureLayer = new FeatureLayer(featureTable);
 
-      // add the layer to the ArcGISMap
+      // add the feature layer to the ArcGISMap
```
Comment (tschie) [inline editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java]:
```suggestion
      // add the feature layer to the map
```

File: editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java [hunk 10]
```diff
@@ -118,21 +114,30 @@ public void start(Stage stage) {
...
       map.getOperationalLayers().add(featureLayer);
 
+      // add a done loading listener, with a runnable that gets triggered asynchronously when the feature layer has loaded
```
Comment (tschie) [inline editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java]: too verbose

File: editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java [hunk 11]
```diff
@@ -153,12 +158,12 @@ public void start(Stage stage) {
...
                 // get selected feature
                 if (element instanceof ArcGISFeature) {
-                  selected = (ArcGISFeature) element;
-                  featureLayer.selectFeature(selected);
-                  selected.loadAsync();
-                  selected.addDoneLoadingListener(() -> {
-                    if (selected.getLoadStatus() == LoadStatus.LOADED) {
-                      comboBox.getSelectionModel().select((String) selected.getAttributes().get("typdamage"));
+                  selectedFeature = (ArcGISFeature) element;
+                  featureLayer.selectFeature(selectedFeature);
```
Comment (tschie) [inline editing/update-attributes/src/main/java/com/esri/samples/update_attributes/UpdateAttributesSample.java]: Perhaps `identifiedFeature` would be a better name? Looks odd to select the selected feature.

File: image_layers/map-image-layer-sublayer-visibility/src/main/java/com/esri/samples/map_image_layer_sublayer_visibility/MapImageLayerSublayerVisibilitySample.java [hunk 12]
```diff
@@ -75,27 +73,39 @@ public void start(Stage stage) {
...
-      imageLayer.setOpacity(0.7f);
+      ArcGISMapImageLayer imageLayer = new ArcGISMapImageLayer("https://sampleserver6.arcgisonline.com/arcgis/rest/services/SampleWorldCities/MapServer");
 
       // add world cities layers as ArcGISMap operational layer
       map.getOperationalLayers().add(imageLayer);
 
-      // set the ArcGISMap to be displayed in this view
-      mapView = new MapView();
-      mapView.setMap(map);
+      // set the image layer's opacity
```
Comment (tschie) [inline image_layers/map-image-layer-sublayer-visibility/src/main/java/com/esri/samples/map_image_layer_sublayer_visibility/MapImageLayerSublayerVisibilitySample.java]: I realize there wasn't a comment before, but we should really have one for why we changed the opacity

File: image_layers/map-image-layer-sublayer-visibility/src/main/java/com/esri/samples/map_image_layer_sublayer_visibility/MapImageLayerSublayerVisibilitySample.java [hunk 13]
```diff
@@ -75,27 +73,39 @@ public void start(Stage stage) {
...
+      imageLayer.addDoneLoadingListener(() -> {
+        if (imageLayer.getLoadStatus() == LoadStatus.LOADED) {
+
+          // get the layers from the ArcGISMap image layer
+          SublayerList layers = imageLayer.getSublayers();
 
-      // get the layers from the ArcGISMap image layer
-      SublayerList layers = imageLayer.getSublayers();
+          // handle sub layer selection
+          citiesBox.selectedProperty().addListener(e -> layers.get(0).setVisible(citiesBox.isSelected()));
```
Comment (tschie) [inline image_layers/map-image-layer-sublayer-visibility/src/main/java/com/esri/samples/map_image_layer_sublayer_visibility/MapImageLayerSublayerVisibilitySample.java:100]: if only we had a visible property to bind to.... @mbcoder 😜

File: raster/raster-rendering-rule/src/main/java/com/esri/samples/raster_rendering_rule/RasterRenderingRuleSample.java [hunk 14]
```diff
@@ -107,11 +107,12 @@ public RenderingRuleInfo fromString(String string) {
...
       // create an Image Service Raster as a raster layer and add to map
       final String ImageServiceRasterUri = "https://sampleserver6.arcgisonline.com/arcgis/rest/services/CharlotteLAS/ImageServer";
-      final ImageServiceRaster imageServiceRaster = new ImageServiceRaster(ImageServiceRasterUri);
-      final RasterLayer imageRasterLayer = new RasterLayer(imageServiceRaster);
+      ImageServiceRaster imageServiceRaster = new ImageServiceRaster(ImageServiceRasterUri);
+      RasterLayer imageRasterLayer = new RasterLayer(imageServiceRaster);
       map.getOperationalLayers().add(imageRasterLayer);
 
-      // add event listener to loading of Image Service Raster and wait until loaded
+      // add a done loading listener, with a runnable that gets triggered asynchronously when the feature layer has loaded
```
Comment (tschie) [inline raster/raster-rendering-rule/src/main/java/com/esri/samples/raster_rendering_rule/RasterRenderingRuleSample.java]: too verbose

**Review by tschie — APPROVED** [review]
Looks good. Just a few minor things
